svcdiscovery/plugin/svcdiscovery/plugin.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2020 SUSE
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package svcdiscovery
+
+import (
+	"context"
+	"net"
+
+	"github.com/coredns/coredns/plugin"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+)
+
+// ServiceDiscovery is a CoreDNS plugin for handling Cloud Foundry App Service
+// Discovery.
+type ServiceDiscovery struct {
+	Next plugin.Handler
+
+	log       clog.P
+	sdcClient *SDCClient
+	ttl       uint32
+}
+
+// Name returns the plugin name.
+func (sd *ServiceDiscovery) Name() string { return pluginName }
+
+// ServeDNS handles domain name requests for Cloud Foundry App Service Discovery
+// by resolving internal domain names.
+func (sd *ServiceDiscovery) ServeDNS(
+	ctx context.Context,
+	res dns.ResponseWriter,
+	req *dns.Msg,
+) (int, error) {
+	state := request.Request{W: res, Req: req}
+	qclass := state.QClass()
+	qtype := state.QType()
+	qname := state.Name()
+
+	if qclass == dns.ClassINET && (qtype == dns.TypeA || qtype == dns.TypeAAAA) {
+		ips, err := sd.sdcClient.Discover(qname)
+		if err != nil {
+			sd.log.Error(err)
+			return dns.RcodeServerFailure, err
```

Comment:
If I understand things correctly, we're return a server failure for DNS calls that we don't cover (say, looking up `www.google.com`) that happens to error out? That seems… odd.  But maybe that's okay, if we expect that to be very rare?

Reply:
No. An error would only occur on connection failure (or something else other than not finding google.com).
When the service discovery controller doesn't have any routes to respond, we get a success response, but no IPs.
If you continue reading, you'll see that we only set out a response writer if the list of IPs is not empty. Any other case would just be forwarded to the upstream server (configured on the Corefile on KubeCF).